### ISSUE TITLE
Experimental decorators do not get access to the initial value when using auto-accessors

### ISSUE THREAD

**Issue by justinfagnani** [microsoft/TypeScript#55669]
### 🔎 Search Terms

In order to give a good out-of-the-box experience with our library that vends decorators, we are trying to make a single set of decorators that works under the following configurations:

1. Experimental decorators on class fields (with `useDefineForClassFields: false`)
2. Standard decorators on auto-accessors
3. Experimental decorators on auto-accessors

Case (3) may seem unusual at first, but it's very important for incremental migration from experimental decorators to standard decorators. The goal is to enable experimental decorators to be _callsite_ compatible with standard decorators so that callsites and files can be upgraded one-at-a-time without changing the tsconfig, and after the callsites are updated `experimentalDecorators` can be turned off.

This mostly works except that in case (3) the decorators cannot get access to the initial field value because it's set directly to the auto-accessor backing store and skips the setter. In case (1) the initial value goes through the setter, in case (2) it goes through the `init()` callback, but in case (3) we can't see it at all.

Case (3) is actually important for multiple reasons:
- Incremental migration as mentioned
- It enables one version of decorator docs (aside from talking about configs) and decorator-using code samples through our docs and example repositories.
- It enables projects to start with or migrate to standard decorators and change back to experimental decorators in case the size penalty of standard decorators is too much (we're seeing a ~50% code size penalty on porting a library to standard decorators above the size with experimental decorators) by just changing the tsconfig.

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about decorators

### ⏯ Playground Link

https://www.typescriptlang.org/play?experimentalDecorators=true&target=9#code/MYewdgzgLgBFCm0YF4YAoBQMYAcBOIUIA8ngCKLB4CWOReAXDMQEYBW8wUANFjGAEMAtvCYAFAjnh4oATwDS8Wb2wATSjTohGMCSCkzZFCFVr0MASiYCwslAD4YAbz554UAK54wzvtgDm7mgWvthhMKCQIAA28AB00SD+aADkgVApFgDcfmFunt4w6iaa9HHpAIRxwALR0WhQABbUENm5AL4qYRBBAG5MHmAA1mAgAO5gIS7h2JEQMfGJySk9GdwwvW0zRRpm2nGrVTV1Dc0Q65s54e18NzcYwNECEBAwAIKhMAACCEgCwMBEPM8DAAGYgEAoGApFgCPApHL3AD0SJgAFEAB5SLjwVQRcDzWJMABEq2J62JsLwxIeBNgAihYHgY3ewRyGBRMCWEBJZIpsIAXjSBHFwZDUDCBAKERggA

### 💻 Code

```ts
const test = (
  protoOrDescriptor: Object,
  name: PropertyKey,
  descriptor: PropertyDescriptor
): any => {
  return {
    get() {
      console.log('get');
      return descriptor.get!.call(this);
    },
    set(v: unknown) {
      console.log('set', v);
      descriptor.set!.call(this, v);
    }
  }
}

class A {
  @test accessor foo = 'bar';
}

// Expected console: "set", "bar"
const a = new A();

// logs: "set", "baz"
a.foo = 'baz';

```

### 🙁 Actual behavior

The setter isn't called, since the output initializes the private storage directly:

```ts
class A {
    #foo_accessor_storage = 'bar';
    get foo() { return this.#foo_accessor_storage; }
    set foo(value) { this.#foo_accessor_storage = value; }
}
```

### 🙂 Expected behavior

The setter is called. Initializers for decorated accessors under experimental decorators are called as statements in the constructor:

```ts
class A {
    #foo_accessor_storage;
    get foo() { return this.#foo_accessor_storage; }
    set foo(value) { this.#foo_accessor_storage = value; }
    constructor() {
      this.foo = 'bar';
    }
}
```

**Comment by trusktr** [comment]
If the new decorators are being designed to be callsite-compatible, why is there a need to support both at the same time? Can't users just update a version number (if not also an import path) and have it just work?

I think had Lit chosen `useDefineForClassFields: true` from the beginning, it would be in the best shape for a version update to simply work.

Elements already have `@element` on them, so you can drop `accessor` and do this (which I've been doing and have not had performance issues):

In other words, if you aim to make it work like this on Legacy Decorators without accessing the prototype in the decorators and with `useDefineForClassFields: true`,

```js
@element('my-el')
class MyEl extends LitElement {
  @property() foo = 123
  @property() bar = 123
  @property() baz = 123
  @property() lorem = 123
  @property() ipsum = 123
}
```

instead of this,

```js
@element('my-el')
class MyEl extends LitElement {
  @property() accessor foo = 123
  @property() accessor bar = 123
  @property() accessor baz = 123
  @property() accessor lorem = 123
  @property() accessor ipsum = 123
}
```

then Lit upadtes to Standard Decorators, Lit users will be able to updgrade without any problems at all (after they fix issues they might unfortunately encounter due to `useDefineForClassFields: false`), and they'll enjoy writing less-verbose properties without `accessor`.

Without `accessor` we convert from own property to getter/setter in `@element`. It won't work without `@element`, is the tradeoff, but I haven't worked in a Lit code base that didn't use both `@property` and `@element`.

**Comment by justinfagnani** [comment]
> If the new decorators are being designed to be callsite-compatible, why is there a need to support both at the same time? Can't users just update a version number (if not also an import path) and have it just work?

No, because the abilities are different. Experimental decorators on auto-accessors can't see the initial value and so can't run them through the logic that standard decorators on accessors can and experimental decorators on "fields" can.

> I think had Lit chosen `useDefineForClassFields: true` from the beginning, it would be in the best shape for a version update to simply work.

Not only did `useDefineForClassFields: true` not exist when Lit was first built, it's not possible to make decorators to work with spec-compliant fields in an acceptable way - this is the whole reason for auto-accessors to exist in the first place.

I don't think it's useful or relevant to litigate Lit's designs decisions that were made before the design of standard decorators because the issue I raised here applies to decorator usage today without any prior decisions.

I know you don't like `accessor`, but I'm not interested in the performance penalty of deleting instance properties of classes to keep class fields from shadowing accessors. You can advocate that approach to those who will listen [here](https://github.com/tc39/proposal-decorators/issues/507#issuecomment-1688450718) without bringing it into unlreated issues. `accessor` is the right way to use decorators and accessors, full stop.

**Comment by trusktr** [comment]
I'm just trying to be constructive by providing suggestions (as a Lit user).

> I'm not interested in the performance penalty of deleting instance properties of classes to keep class fields from shadowing accessors

My suggestion was slightly different: replacing own properties with own getters/setters.

> `accessor` is the right way to use decorators and accessors, full stop.

I accept your opinion. Some people will like to avoid `accessor` everywhere, and will take the negligible cost.

It's like comparing the speed of Lit vs Solid/React/Vue/etc: the difference in practice won't matter much.

I am 100% confident shipping own getters/setters will have no impact on rocket UI functionilty and performance. I've used this approach for 1.5 years already, and it's been great.

**Comment by justinfagnani** [comment]
@andrewbranch Is there any chance of this getting looked at? It's still assigned to rbuckton, who I don't think is on the TypeScript team anymore.

A solution would be to run the initial value through the setter, just like field initializers are.

Fixing this would let us remove workaround code in Lit.

**Comment by RyanCavanaugh** [comment]
@justinfagnani isn't this enormously likely to break someone who was depending on the setter *not* running?

**Comment by justinfagnani** [comment]
@RyanCavanaugh 

This is just a guess, but no. Experimental decorators + auto-accessors has to be an extremely rare outside of Lit migrations. For most people, if they're using auto-accessors, they'll be using standard decorators. I've tested our decorators for this case (by triggering our accessor decoration path for fields, and running on fields with initializers) and our decorators work in this combination.

But I get that you probably don't want to invest time into experimental decorators and take any risk of dealing with unintended breaks. And now that I think about it, we'd rather just remove all of our experimental decorator code anyway. I suppose we can close this if you're going to deprecate experimental decorators soon.

**Comment by RyanCavanaugh** [comment]
We don't have any reasonable path toward deprecating experimental decorators (way too much usage in ways that can't migrate to standard), but yeah we don't have a huge interest in unintentionally breaking folks or expanding the configuration space for it either
